manager/apis/app/v1alpha1/plotter_types.go

```diff
@@ -4,21 +4,306 @@
+	// +required
+	AssetID string `json:"assetId"`
+}
+
+// StepParameters holds the parameters to the module
+// that is deployed in this step
+type StepParameters struct {
+	// +optional
+	Source *StepSource `json:"source,omitempty"`
+
+	// +optional
+	Sink *StepSink `json:"sink,omitempty"`
+
+	// +optional
+	API *API `json:"api,omitempty"`
+
+	// Actions are the data transformations that the module supports
+	// +optional
+	Actions []serde.Arbitrary `json:"action,omitempty"`
+}
+
+// DataFlowStep contains details on a single data flow step
+type DataFlowStep struct {
+	// Name of the step
+	// +required
+	Name string `json:"name"`
+
+	// Name of the cluster this step is executed on
+	// +required
+	Cluster string `json:"cluster"`
+
+	// Template is the name of the template to execute the step
+	// The full details of the template can be extracted from Plotter.spec.templates
+	// list field.
+	// +required
+	Template string `json:"template"`
+
+	// +optional
+	Parameters *StepParameters `json:"parameters,omitempty"`
+}
+
+// +kubebuilder:validation:Type=array
+type SequentialSteps struct {
+	// Step contains details of a single data flow step: the module to deploy and its inputs and outputs.
+	// The execution of the steps can be parallel or serial and is determined according to the following:
+	// single dash => run after previous step
+	// double dash => run in parallel with previous step
+	// +required
+	Steps []DataFlowStep `json:"-" protobuf:"bytes,1,rep,name=steps"`
+}
+
+// Step is an anonymous list inside of SequentialSteps (i.e. it does not have a key), so it needs its own
+// custom Unmarshaller
+func (p *SequentialSteps) UnmarshalJSON(value []byte) error {
+	// Since we are writing a custom unmarshaller, we have to enforce the "DisallowUnknownFields" requirement manually.
+
+	// First, get a generic representation of the contents
+	var candidate []map[string]interface{}
+	err := json.Unmarshal(value, &candidate)
+	if err != nil {
+		return err
+	}
+
+	// Generate a list of all the available JSON fields of the Step struct
+	availableFields := map[string]bool{}
+	reflectType := reflect.TypeOf(DataFlowStep{})
+	for i := 0; i < reflectType.NumField(); i++ {
+		cleanString := strings.ReplaceAll(reflectType.Field(i).Tag.Get("json"), ",omitempty", "")
+		availableFields[cleanString] = true
+	}
+
+	// Enforce that no unknown fields are present
+	for _, step := range candidate {
+		for key := range step {
+			if _, ok := availableFields[key]; !ok {
+				return fmt.Errorf(`json: unknown field "%s"`, key)
+			}
+		}
+	}
+
+	// Finally, attempt to fully unmarshal the struct
+	err = json.Unmarshal(value, &p.Steps)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p SequentialSteps) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.Steps)
+}
+
+// SubFlowTrigger indicates the trigger for this subflow
+// +kubebuilder:validation:Enum=read;write;copy
+type SubFlowTrigger string
+
+const (
```

Comment:
This is not what I had in mind for triggers.  Triggers are workload, init, timer, event.  They are not specific to a type of data flow.

Reply:
I fixed that.